### DISCUSSION TITLE
Does meshopt_optimizeVertexCache, etc. preserve index order?

### DISCUSSION THREAD

**Discussion by coffeenotfound** [zeux/meshoptimizer#576] [General]
I'm writing a visiblity renderer and for that I need efficient triangle ids. SV_PrimitiveID is very slow so I'm building my own mesh optimizer that injects triangle id's as vertex attributes with flat interpolation

For this I would run meshopt_optimizeVertexCache first, then my own pass and then meshopt_optimizeVertexCache again to reorder the vertices I put at the end of the list to a more sensible location.

So the important question is, does meshopt_optimizeVertexCache preserve the order of indices within a triangle? Or are the no guarentees whatsoever?

**Comment by zeux** [comment]
While there is no formal guarantee that the indices in a triangle maintain their order beyond preserving winding, in practice both vertex cache optimization algorithms as well as most other algorithms with the exception of `meshopt_encodeIndexBuffer` maintain the order and it's very unlikely to change as it hasn't changed in years and there's no real incentive to do so.

**Comment by coffeenotfound** [comment]
Thanks!
